[SiteIsolation] Implement basic "descendant frames" focus advancing across site isolated frames

beidson opened the pull request:
#### 8b2d5e03d8b976c2712356a30fdda4d32132b123
<pre>
[SiteIsolation] Implement basic &quot;descendant frames&quot; focus advancing across site isolated frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=295684">https://bugs.webkit.org/show_bug.cgi?id=295684</a>
&lt;<a href="https://rdar.apple.com/problem/155493711">rdar://problem/155493711</a>&gt;

Reviewed by Alex Christensen.

This implements advancing focus across descendant isolated frames.
e.g.

A
|
-&gt;B
  |
  -&gt;C

Where B and C are different origins from A, and therefore in other processes, focus advancing will
cross process boundaries and finally relinquish to the browser chrome at the end, all before starting
over again up at A.

Known to still not work is sibling isolated frames.
e.g.

  A
  |
&lt;-.-&gt;
|   |
B   C

In this arrangement, advancing focus will move from A to B, then relinquish to the browser chrome
instead of continuing in C.

Making that work is the next step after this patch.

* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::findAndFocusElementStartingWithLocalFrame):
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::findFocusableElementAcrossFocusScope):
* Source/WebCore/page/FocusController.h:

* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::findFocusableElementDescendingIntoRemoteFrame):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::makeWebViewAndDelegates):
(TestWebKitAPI::(SiteIsolation, AdvanceFocusAcrossFrames)):

* Tools/TestWebKitAPI/cocoa/TestUIDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestUIDelegate.mm:
(-[TestUIDelegate _webView:takeFocus:]):

Synchronous access to the frame tree was useful in an earlier iteration of this test, and seems
like it is generally useful to TestWebKitAPI clients, so I&apos;m leaving it in.
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView frameTree]):

Canonical link: <a href="https://commits.webkit.org/297419@main">https://commits.webkit.org/297419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5f0e43712d896291709f9651e468ca3120a351a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111722 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117754 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113684 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32076 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39973 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84887 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114669 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25610 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100555 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65325 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/24945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18687 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61586 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94997 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18759 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121005 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38768 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/28821 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93778 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39148 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96810 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93600 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23861 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38751 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/16543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34798 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38662 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44153 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38316 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41644 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
```diff
@@ -1244,6 +1244,21 @@ - (void)focus
+        isDone = true;
+    }];
+    TestWebKitAPI::Util::run(&isDone);
+
```
nit: extra newline.  Also, we don't need isDone to be separate from result.  We could just spinRunLoop until result is non-null.

beidson replied:
There's multiple examples I can find of a RetainPtr<> being the actual result and a separate `done` variable being required.
I tried removing the `done` and using the smart pointer directly instead and that wouldn't build.

achristensen07 replied:
It's not important, but to build that you would need something like this instead of a call to Util::run
while (!result)
    TestWebKitAPI::Util::spinRunLoop();

This is fine as it is, though.

achristensen07 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -523,6 +524,61 @@ void FocusController::setFocusedInternal(bool focused)
 
+FocusableElementSearchResult FocusController::findAndFocusElementStartingWithLocalFrame(FocusDirection direction, const FocusEventData& focusEventData, LocalFrame& frame)
+{
+    Ref protectedFrame = frame;
```
This shouldn't be necessary.  The caller should ensure lifetime of parameters.

achristensen07 replied:
If I understand correctly, this pattern is contrary to our safer-cpp patterns.  This should be removed.  The caller ensures the lifetime of the LocalFrame.

achristensen07 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -523,6 +524,61 @@ void FocusController::setFocusedInternal(bool focused)
+        return findResult;
+    }
+
+    if (RefPtr owner = dynamicDowncast<HTMLFrameOwnerElement>(*findResult.element); owner && (!is<HTMLPlugInElement>(*findResult.element) || !findResult.element->isKeyboardFocusable(focusEventData))) {
```
I don't think I understand why the HTMLPlugInElement check is here.

beidson replied:
It exists in other code (referenced below...)

achristensen07 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -523,6 +524,61 @@ void FocusController::setFocusedInternal(bool focused)
+    if (frame.settings().caretBrowsingEnabled()) {
+        VisibleSelection newSelection(firstPositionInOrBeforeNode(findResult.element.get()), Affinity::Downstream);
+        if (frame.selection().shouldChangeSelection(newSelection)) {
+            AXTextStateChangeIntent intent(AXTextStateChangeTypeSelectionMove, AXTextSelection { AXTextSelectionDirectionDiscontiguous, AXTextSelectionGranularityUnknown, true });
```
I don't understand where this code is coming from.  Should it be shared with something?

beidson replied:
Trying to refactor where this came from (which was elsewhere in FocusController) made uglier code.

I will try again, since it led to two confusions.

achristensen07 commented on Source/WebCore/page/FocusController.cpp:
```diff
@@ -610,58 +624,79 @@ bool FocusController::relinquishFocusToChrome(FocusDirection direction)
+FocusableElementSearchResult FocusController::findAndFocusElementInDocumentOrderStartingWithFrame(Ref<LocalFrame> frame, RefPtr<Node> scopeNode, RefPtr<Node> startingNode, FocusDirection direction, const FocusEventData& focusEventData, InitialFocus initialFocus, ContinuingRemoteSearch continuingRemoteSearch)
+{
+    RefPtr document = frame->document();
+    RELEASE_ASSERT(document);
```
We could also just debug assert and return an empty result in this case.

achristensen07 commented on Source/WebCore/page/FocusController.h:
```diff
@@ -95,13 +95,20 @@ class FocusController final : public CanMakeCheckedPtr<FocusController> {
     void setFocusedInternal(bool);
     void setIsVisibleAndActiveInternal(bool);
 
+    enum class InitialFocus { No, Yes };
```
: bool

achristensen07 commented on Source/WebCore/page/FocusController.h:
```diff
@@ -95,13 +95,20 @@ class FocusController final : public CanMakeCheckedPtr<FocusController> {
     void setIsVisibleAndActiveInternal(bool);
 
+    enum class InitialFocus { No, Yes };
+    enum class ContinuingRemoteSearch { No, Yes };
```
: bool

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -1538,10 +1538,21 @@ std::optional<ResourceResponse> WebFrame::resourceResponseForURL(const URL& url)
+        if (RefPtr localFrame = dynamicDowncast<LocalFrame>(m_coreFrame.get())) {
+            if (RefPtr page = localFrame->page()) {
+                auto result = page->focusController().findAndFocusElementStartingWithLocalFrame(direction, focusEventData, *localFrame);
+                if (result.element)
+                    foundElementInRemoteFrame = WebCore::FoundElementInRemoteFrame::Yes;
```
This is confusing.  How can we find an element in a RemoteFrame?  Wouldn't the result.element be in a different process in that case?

beidson replied:
Note - The answer is not *the element*.
But rather *was an element found*

That is already baked in to the IPC message itself. Because in the larger overall work here, whether or not an element was found remotely will have impact on what the next step is in the focus operation.
Namely, the originating process is who decides to give up and relinquish focus to the chrome, and this answer informs that.

(We're simply not there yet)

beidson committed:
[SiteIsolation] Implement basic "descendant frames" focus advancing across site isolated frames
https://bugs.webkit.org/show_bug.cgi?id=295684
<rdar://problem/155493711>

Reviewed by Alex Christensen.

This implements advancing focus across descendant isolated frames.
e.g.

A
|
->B
  |
  ->C

Where B and C are different origins from A, and therefore in other processes, focus advancing will
cross process boundaries and finally relinquish to the browser chrome at the end, all before starting
over again up at A.

Known to still not work is sibling isolated frames.
e.g.

  A
  |
<-.->
|   |
B   C

In this arrangement, advancing focus will move from A to B, then relinquish to the browser chrome
instead of continuing in C.

Making that work is the next step after this patch.

* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::findAndFocusElementStartingWithLocalFrame):
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::findFocusableElementAcrossFocusScope):
* Source/WebCore/page/FocusController.h:

* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::findFocusableElementDescendingIntoRemoteFrame):

* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::makeWebViewAndDelegates):
(TestWebKitAPI::(SiteIsolation, AdvanceFocusAcrossFrames)):

* Tools/TestWebKitAPI/cocoa/TestUIDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestUIDelegate.mm:
(-[TestUIDelegate _webView:takeFocus:]):

Synchronous access to the frame tree was useful in an earlier iteration of this test, and seems
like it is generally useful to TestWebKitAPI clients, so I'm leaving it in.
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[TestWKWebView frameTree]):

Canonical link: https://commits.webkit.org/297419@main
